### PR TITLE
Worklist generalization

### PR DESCRIPTION
**Pull request by thorulf4**
This attempts to make a work list generalization technique to abstract the worklist aspect away from the core analysis code.

So far this generalization does not allow any changes to transfer functions
But it allows custom analysis code with ability to pass additional parameters to said code. This is done with the Analyzer classes.
It also allows custom node container types, needed for runtime sized arrays. This is done through the NodeUploader struct

### CODE REVIEW

File: src/cuda/cuda_worklist/analysis.cu [hunk 1]
```diff
@@ -1,153 +1,31 @@
...
+void cuda_worklist::execute_analysis(DynamicArray<Node>& nodes, std::vector<Transfer>& transfers, const std::set<int>& taint_sources){
+    Analyzer analyzer;
 
-    cuda_free(dev_nodes);
-    cuda_free(dev_transfers);
-    cuda_free(dev_worklists);
+    worklist::NodeUploader<Node*> uploader;
+    uploader.dev_nodes = (void**)&uploader.container;
+    
+    worklist::execute_some_analysis(analyzer, nodes, uploader, &transfers[0], transfers.size(), taint_sources);
```
Comment (NicklasHugoy) [inline src/cuda/cuda_worklist/analysis.cu:30]: Would be nice with another name for this function

File: src/main.cpp [hunk 2]
```diff
@@ -92,8 +92,10 @@ std::vector<StatefulNode<std::set<std::string>>> bit_cuda_worklist_analysis(ScTr
                 reduce_variables, program.entryNodes);
     auto transformer = time_func<CudaTransformer<cuda_worklist::Node>>("Gpu structure transformation: ", 
                 transform_cuda_worklist, program.nodes);
+    // time_func("Least fixed point algorithm: ",
+    //         cuda_worklist::execute_analysis, &transformer.nodes[0], transformer.nodes.size(), &*transformer.transfer_functions.begin(), transformer.transfer_functions.size(), transformer.taint_sources);
```
Comment (NicklasHugoy) [inline src/main.cpp:96]: Commented code

File: src/cuda/worklist/analysis.cuh [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+#include <iostream>
+#include <stdio.h>
+#include <vector>
+#include <array>
+#include "../cuda_common.cuh"
+#include <type_traits>
+#include <timing.h>
+
+
+namespace worklist{
```
Comment (NicklasHugoy) [inline src/cuda/worklist/analysis.cuh:13]: maybe call the file `worklist.cuh`?

**Review by ZcortezZ — APPROVED** [review]

**Review by NicklasHugoy — CHANGES_REQUESTED** [review]

File: src/cuda/cuda_common.cu [hunk 4]
```diff
@@ -37,3 +37,10 @@ void init_gpu(){
     cuda_free(x);
 }
 
+void cuda_allocate_memory(void** devPtr, size_t size){
+    auto cudaStatus = cudaMalloc(devPtr, size);
+    if (cudaStatus != cudaSuccess) {
+        fprintf(stderr, "cudaMalloc failed: %s\n", cudaGetErrorString(cudaStatus));
+        exit(1);
+    }
+}
```
[the remark this answers is not in the document]
Comment (thorulf4) [reply src/cuda/cuda_common.cu:46]: This is only used for one case where the templated option isnt suited

**Review by ZcortezZ — APPROVED** [review]

**Review by NicklasHugoy — APPROVED** [review]

**Pull request merged**